File: book_this-is-coding-test-with-python/part2/chap3-greedy/p2/answer.js
// import readline from "readline";

// const rl = readline.createInterface({
//   input: process.stdin,
//   output: process.stdout,
// });

// let lines = [];
// let lineLimit = 0;
// let curLine = 0;

// rl.on("line", (line) => {
//   lines.push(line);

//   if (curLine === 0) {
//     lineLimit = +line.split(" ")[0] + 1;
//   }

//   curLine++;
//   if (curLine === lineLimit) {
//     rl.close();
//   }
// }).on("close", () => {
//   let data = [];
//   for (let i = 1; i <= lineLimit - 1; i++) {
//     data.push(lines[i].split(" ").map((num) => parseInt(num)));
//   }

//   let result = 0;
//   for (const row of data) {
//     const minValue = Math.min(...row);
//     if (minValue > result) result = minValue;
//   }

//   console.log(result);
// });

function solution(n, m, data) {
  let result = 0;
  for (const row of data) {
    const minValue = Math.min(...row);
    if (minValue > result) result = minValue;
  }

  return result;
}

module.exports = solution;

File: book_this-is-coding-test-with-python/part2/chap8-dp/p1/answer.py
d = dict()


def find_min_count(x):
    if x == 1:
        d[x] = 0
        return 0

    if d.get(x):
        return d[x]

    calc_list = [find_min_count(x - 1)]
    if x % 2 == 0:
        calc_list.append(find_min_count(x / 2))
    if x % 3 == 0:
        calc_list.append(find_min_count(x / 3))
    if x % 5 == 0:
        calc_list.append(find_min_count(x / 5))

    d[x] = 1 + min(calc_list)
    return d[x]


x = int(input())

print(find_min_count(x))
